technical-docs/K8SSANDRA-OPERATOR-DESIGN.md

```diff
@@ -0,0 +1,410 @@
+    clusterName: k8ssandra
+    auth:
+      enabled: true
+    datacenters:
+    - name: dc1
+      size: 3
+  reaper:
+    enabled: true
+    serverConfig:
+      jmxUserSecretName: my-custom-secret
+      cassandraBackend:
+        replication:
+          networkTopology:
+            dc1: 5         
+  stargate: 
+    enabled: true  
+    replicas: 3
+```
+
+### CassandraCluster
+
+```
+apiVersion: k8ssandra.io/v1alpha1
+kind: CassandraCluster
+metadata:
+  name: k8ssandra
+spec:
+  clusterName: k8ssandra
+  auth:
+    enabled: true
+  datacenters:
+  - name: dc1
+    size: 3
+```
+
+The controller will create a CassandraDatacenter object for each entry in the datacenters array. Similar to what we have done in the Helm charts, you will be able to specify things at cluster level and override them at the DC level. You will also be able to specify properties at the DC level without specifying them at the cluster level.
+
+Multi-cluster is out of scope, but I want to point out that I think this is where we introduce it. Consider the following example:
+
+```
+apiVersion: k8ssandra.io/v1alpha1
+kind: CassandraCluster
+metadata:
+  name: k8ssandra
+spec:
+  clusterName: k8ssandra
+  auth:
+    enabled: true
+  datacenters:
+  - name: dc1
+    size: 3
+  - name: dc2
+    k8s:
+      kubeconfigSecret: my-other-cluster
+      namespace: dev
+    size: 3
+```
+
+`kubeconfigSecret` specifies the name of a secret that contains the kubeconfig file needed to access a remote cluster. 
+
+`namespace` indicates in which namespace the DC should be created.
+
+### CassandraDatacenter
+
+This is defined in the [cass-operator](https://github.com/k8ssandra/cass-operator) repo. The controller runs in its own Deployment.
+
+### Reaper
+
+```
+apiVersion k8ssandra.io/v1alpha1
+kind: Reaper
+metadata:
+  name: k8ssandra
+spec:
+  serverConfig:
+    storageType: cassandra
+    jmxUserSecretName: k8ssandra
+    cassandraBackend:
+      cassandraDatacenter
+        name: dc1
+      replication:
+        networkTopology:
+          dc1: 3
+```
+
+This is defined in the [reaper-operator](https://github.com/k8ssandra/reaper-operator) repo. It is possible today to use reaper-operator on its own without k8ssandra. In that regard, living in its own repo makes sense. There is a good bit of overhead for each repo that we need to maintain. In practice, I do not think reaper-operator will be used a whole lot outside of k8ssandra. For these reasons I am in favor of consolidating with the k8ssandra-operator repo.
+
+From a technical standpoint, consolidation means that the reaper-operator controllers would run in the k8ssandra-operator binary. It would reduce overhead in that it would be able to use the same cache as other controllers.
+
+### Stargate
+
+This does not yet exist. Similar to reaper-operator, one could certainly make the case that a separate stargate-operator project makes sense. I am more on the fence about where this should live than I am with reaper-operator.
```

Comment:
It's probably a fair question to ask if an operator for Stargate should live in K8ssandra at all, or if there is external value in the Stargate operator residing with Stargate in its org.  Certainly complicates things, and perhaps should start here, but we should definitely have that conversation.

Reply:
If Stargate is separate then it makes medusa / reaper ingestion quite odd. I'd say either all of them, or none of them.

Reply:
To start everything with the exception of cass-operator should live in-tree.